pkg/sst/sst.go

```diff
@@ -395,3 +423,244 @@ func DisableBF(pkgs ...int) error {
+		req = setBit(0, 1)
+
+		if info.CPPriority > 0 {
+			req = setBit(req, 2)
+		}
+	}
+
+	if _, err := sendMboxCmd(cpu, CONFIG_CLOS, CLOS_PM_QOS_CONFIG, param, req); err != nil {
+		return fmt.Errorf("failed to set SST-CP status: %v", err)
+	}
+
+	return nil
+}
+
+func enableCP(info *SstPackageInfo, cpu utils.ID) (uint32, error) {
+	if err := writeClosPmQosConfig(info, cpu, true); err != nil {
+		return 0, fmt.Errorf("Cannot set Clos status: %v", err)
+	}
+
+	return writePMConfig(info, cpu, true)
+}
+
+func disableCP(info *SstPackageInfo, cpu utils.ID) (uint32, error) {
+	if err := writeClosPmQosConfig(info, cpu, false); err != nil {
+		return 0, fmt.Errorf("Cannot set Clos status: %v", err)
+	}
+
+	return writePMConfig(info, cpu, false)
+}
+
+func setDefaultClosParam(info *SstPackageInfo, cpu utils.ID) error {
+	defaultConfig := &SstClosInfo{MaxFreq: 255}
+
+	for clos := 0; clos < 4; clos++ {
+		if err := saveClos(defaultConfig, cpu, clos); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func assignCPU2Clos(info *SstPackageInfo, clos int) error {
+	sstlog.Debugf("Setting Clos %d for cpus %v\n", clos, info.ClosCPUInfo[clos].Members())
+
+	for _, cpu := range info.ClosCPUInfo[clos].Members() {
+		if err := associate2Clos(cpu, clos); err != nil {
+			return fmt.Errorf("failed to associate cpu %d to clos %d: %v", cpu, clos, err)
+		}
+	}
+
+	return nil
+}
+
+// ConfigureCP will allow caller to configure CPUs to various Clos.
+func ConfigureCP(info *SstPackageInfo, priority int, cpu2clos *ClosCPUSet) error {
+	if priority < 0 || priority > 1 {
+		return fmt.Errorf("Invalid CP priority value %d (valid 0 or 1)", priority)
+	}
+
+	if info.ClosCPUInfo == nil {
+		info.ClosCPUInfo = make(map[int]utils.IDSet, len(*cpu2clos))
+	}
+
+	for clos, cpus := range *cpu2clos {
+		info.ClosCPUInfo[clos] = cpus.Clone()
+
+		// Remove the CPU from other Clos if found
+		for i := 0; i < NumClos; i++ {
+			if i == clos {
+				continue
+			}
+
+			for id := range cpus {
+				info.ClosCPUInfo[i].Del(id)
+			}
+		}
+
+		if err := assignCPU2Clos(info, clos); err != nil {
+			return err
+		}
+	}
+
+	info.CPPriority = CPPriorityType(priority)
+
+	return nil
+}
+
+// ClosSetup stores the user supplied Clos information into punit
+func ClosSetup(info *SstPackageInfo, clos int, closInfo *SstClosInfo) error {
+	if clos < 0 || clos >= NumClos {
+		return fmt.Errorf("Invalid Clos value (%d)", clos)
+	}
+
+	info.ClosInfo[clos] = *closInfo
+
+	return saveClos(&info.ClosInfo[clos], info.pkg.cpus[0], clos)
+}
+
+// ResetCPConfig will bring the system to a known state. This means that all
+// CLOS groups are reset to their default values, all package cores are assigned to
+// CLOS group 0 and ordered priority mode is enabled.
+func ResetCPConfig() error {
+	infomap, err := GetPackageInfo()
+	if err != nil {
+		return err
+	}
+
+	for _, info := range infomap {
+		for _, cpu := range info.pkg.cpus {
+			if err := setDefaultClosParam(info, cpu); err != nil {
+				return err
+			}
```

Comment:
Hmm, I don't think you need to do this for every logical cpu, just once per each package. I.e. for `info.pkg.cpus[0]` similar to `ClosSetup()` above